--- a/net/minecraft/server/level/DistanceManager.java
+++ b/net/minecraft/server/level/DistanceManager.java
@@ -177,7 +177,7 @@ public abstract class DistanceManager implements ca.spottedleaf.moonrise.patches
         for (int i = 0; i < size; ++i) {
             final LevelChunk chunk = raw[i];
 
-            action.accept(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunk.getPos()));
+            action.accept(chunk.getPos().longKey); // Leaf - Cache chunk key
         }
         // Paper end - rewrite chunk system
     }
