From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Canvas <canvas@canvasmc.io>
Date: Fri, 31 Jul 2026 11:08:00 +0000
Subject: [PATCH] Virtual thread support for chat executor


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 29ece5fe35571a62580948257a81bde1e9945256..bf8fc02aea6663ba7dee6730c52cb075683c42a9 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -2900,8 +2900,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         }
     }
 
-    public final java.util.concurrent.ExecutorService chatExecutor = java.util.concurrent.Executors.newCachedThreadPool(
-        new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon(true).setNameFormat("Async Chat Thread - #%d").setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(net.minecraft.server.MinecraftServer.LOGGER)).build()); // Paper
+    public final java.util.concurrent.ExecutorService chatExecutor = io.canvasmc.canvas.async.ExecutorFactory.buildChatExecutor(); // Canvas - Virtual thread support for chat executor
     public final ChatDecorator improvedChatDecorator = new io.papermc.paper.adventure.ImprovedChatDecorator(this); // Paper - adventure
 
     public ChatDecorator getChatDecorator() {
