--- a/net/minecraft/util/Util.java
+++ b/net/minecraft/util/Util.java
@@ -98,22 +98,9 @@
     private static final TracingExecutor BACKGROUND_EXECUTOR = makeExecutor("Main", -1); // Paper - Perf: add priority
     private static final TracingExecutor IO_POOL = makeIoExecutor("IO-Worker-", false);
     public static final TracingExecutor DIMENSION_DATA_IO_POOL = makeExtraIoExecutor("Dimension-Data-IO-Worker-"); // Paper - Separate dimension data IO pool
-    // Paper start - Limit download pool size
-    private static final TracingExecutor DOWNLOAD_POOL = new TracingExecutor(Executors.newFixedThreadPool(4, new java.util.concurrent.ThreadFactory() {
-        private final AtomicInteger count = new AtomicInteger();
-
-        @Override
-        public Thread newThread(Runnable run) {
-            Thread ret = new Thread(run);
-            ret.setDaemon(true);
-            ret.setName("Download-" + this.count.getAndIncrement());
-            ret.setUncaughtExceptionHandler((Thread thread, Throwable throwable) -> {
-                LOGGER.error("Uncaught exception in thread " + thread.getName(), throwable);
-            });
-            return ret;
-        }
-    }));
-    // Paper end - Limit download pool size
+    private static class DownloadPoolHolder {
+        private static final TracingExecutor INSTANCE = new TracingExecutor(io.canvasmc.canvas.async.ExecutorFactory.buildDownloadPoolExecutor()); // Canvas - Virtual thread support for download pool
+    }
     private static final DateTimeFormatter FILENAME_DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd_HH.mm.ss", Locale.ROOT);
     public static final int LINEAR_LOOKUP_THRESHOLD = 8;
     private static final Set<String> ALLOWED_UNTRUSTED_LINK_PROTOCOLS = Set.of("http", "https");
@@ -247,7 +234,7 @@
     }
 
     public static TracingExecutor nonCriticalIoPool() {
-        return DOWNLOAD_POOL;
+        return DownloadPoolHolder.INSTANCE;
     }
 
     public static void shutdownExecutors() {
