--- a/net/minecraft/util/ZeroBitStorage.java
+++ b/net/minecraft/util/ZeroBitStorage.java
@@ -5,7 +_,7 @@
 import org.apache.commons.lang3.Validate;
 
 public class ZeroBitStorage implements BitStorage {
-    public static final long[] RAW = new long[0];
+    public static final long[] RAW = io.canvasmc.canvas.util.ArrayConstants.emptyLongArray; // Canvas - reduce array allocations
     private final int size;
 
     public ZeroBitStorage(int size) {
@@ -80,4 +_,6 @@
         return ret;
     }
     // Paper end - block counting
+
+    @Override public <T> void compact(net.minecraft.world.level.chunk.Palette<T> srcPalette, net.minecraft.world.level.chunk.Palette<T> dstPalette, short[] out, net.minecraft.world.level.chunk.PalettedContainer<T> resizeHandler) {}  // Canvas - Lithium - faster chunk serialization
 }
