--- a/net/minecraft/network/CipherBase.java
+++ b/net/minecraft/network/CipherBase.java
@@ -7,8 +_,8 @@
 
 public class CipherBase {
     private final Cipher cipher;
-    private byte[] heapIn = new byte[0];
-    private byte[] heapOut = new byte[0];
+    private byte[] heapIn = io.canvasmc.canvas.util.ArrayConstants.emptyByteArray;
+    private byte[] heapOut = io.canvasmc.canvas.util.ArrayConstants.emptyByteArray;
 
     protected CipherBase(Cipher cipher) {
         this.cipher = cipher;
